buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy

```diff
@@ -96,9 +103,61 @@ class PrecommitTasks {
         }
         Task forbiddenApis = project.tasks.findByName('forbiddenApis')
         forbiddenApis.group = "" // clear group, so this does not show up under verification tasks
+
         return forbiddenApis
     }
 
+    private static Task configureForbiddenApisCli(Project project) {
+        project.configurations.create("forbiddenApisCliJar")
+        project.dependencies {
+            forbiddenApisCliJar 'de.thetaphi:forbiddenapis:2.5'
+        }
+        Task forbiddenApisCli = project.tasks.create('forbiddenApisCli')
+
+        project.sourceSets.forEach { sourceSet ->
+            forbiddenApisCli.dependsOn(
+                project.tasks.create(sourceSet.getTaskName('forbiddenApisCli', null), JavaExec) {
+                    classpath = project.files(
+                            project.configurations.forbiddenApisCliJar,
+                            sourceSet.compileClasspath,
+                            sourceSet.runtimeClasspath
+                    )
+                    main = 'de.thetaphi.forbiddenapis.cli.CliMain'
+                    executable = "${project.runtimeJavaHome}/bin/java"
+                    [
+                            'jdk-unsafe-1.8', 'jdk-deprecated-1.8', 'jdk-non-portable', 'jdk-system-out'
+                    ].forEach { args "-b", it }
+                    [
+                            getForbiddenSignaturesFile(project, 'jdk-signatures'),
+                            getForbiddenSignaturesFile(project, 'es-all-signatures')
+                    ].forEach { args "-f", it }
+                    if (sourceSet.name == 'test') {
+                        args "-f", getForbiddenSignaturesFile(project, 'es-test-signatures')
+                        args "-f", getForbiddenSignaturesFile(project, 'http-signatures')
+                    } else {
+                        args "-f", getForbiddenSignaturesFile(project, 'es-server-signatures')
+                    }
+                    args "--suppressannotation", '**.SuppressForbidden'
+                    dependsOn sourceSet.classesTaskName
+                    doFirst {
+                        // Forbidden APIs expects only existing dirs, and requires at least one
+                        FileCollection existingOutputs = sourceSet.output.classesDirs
+                                .filter { it.exists() }
+                        if (existingOutputs.isEmpty()) {
+                            throw new StopExecutionException("${sourceSet.name} has no outputs")
+                        }
+                        existingOutputs.forEach { args "-d", it }
+                    }
+                }
+            )
+        }
+        return forbiddenApisCli
+    }
+
+    private static File getForbiddenSignaturesFile(Project project, String name) {
+        return project.rootProject.file('buildSrc/src/main/resources/forbidden/' + name + '.txt')
```

Comment:
We can't assume this is accessible, since build-tools is used by external plugin developers. We need to load this from resources and copy to a local file.

Reply:
I wasn't aware we publish `group: 'org.elasticsearch.gradle', name: 'build-tools', version: '6.3.1'` and encourage others to use it.  Is this fully documented or more like plugin authors picked up on it ? I don't think we have documentation on any of the build plugins.  I think I broke this assumption recently with the naming conventions check, I'll take a look and see if we can add some tests for this.

Reply:
Naming conventions check is fine, it looks up the source jar of the relevant class so works as long as the jar is on the file system, the class doesn't have to be. I'll still add a test for this.

Reply:
It's not "documented", but we have encouraged plugin authors to use it since we switched to gradle in 5.0. We have long needed a permanent home for plugin author documentation, but we probably need a broader reorg of the docs to enable that, as there are a few parts for plugin authors to consume eg how to build, how to test, core apis and services available, how to publish, etc.